File: dynamiclayout.php
<?php
/**
 * DynamicLayout PHP API — generate layout JSON from PHP code.
 *
 * Single-file. Zero dependencies (json_encode is built-in).
 * Works on PHP 7.4+.
 *
 * Example:
 *   $page = Layout::begin('My Page');
 *   $page->fieldset('Info')->label('Hello from PHP!')->end();
 *   $page->button('ok', 'OK', 'primary', true);
 *   echo $page->end();
 */

class Layout {
    private string $title;
    private string $uid;
    private array $layout = [];
    private array $actions = [];
    private array $stack = [];
    private int $counter = 0;

    private function __construct(string $title) {
        $this->title = $title;
        $this->uid = 'layout_' . hrtime(true);
        $this->stack[] = &$this->layout;
    }

    public static function begin(string $title): self { return new self($title); }

    private function k(string $prefix): string { return $prefix . '_' . (++$this->counter); }

    private function &top(): array & { return $this->stack[count($this->stack) - 1]; }

    private function add(array $el): self { $this->top()[] = $el; return $this; }

    private function push(array $el): self {
        $el['content'] = []; $this->add($el); $this->stack[] = &$this->top()[count($this->top()) - 1]['content'];
        return $this;
    }

    private function pop(): self { array_pop($this->stack); return $this; }

    // Containers
    public function fieldset(string $title): self { return $this->push(['type' => 'FIELDSET', 'key' => $this->k('fs'), 'title' => $title]); }
    public function end(): self { return $this->pop(); }
    public function row(): self { return $this->push(['type' => 'ROW', 'key' => $this->k('r')]); }
    public function endRow(): self { return $this->pop(); }
    public function col(int $xs = 0, int $sm = 0, int $md = 0, int $lg = 0, int $xl = 0): self {
        $el = ['type' => 'COL', 'key' => $this->k('c')];
        $len = []; if ($xs) $len['xs'] = $xs; if ($sm) $len['sm'] = $sm; if ($md) $len['md'] = $md; if ($lg) $len['lg'] = $lg; if ($xl) $len['xl'] = $xl;
        if ($len) $el['length'] = $len;
        return $this->push($el);
    }
    public function endCol(): self { return $this->pop(); }

    // Display
    public function label(string $text): self { return $this->add(['type' => 'LABEL', 'key' => $this->k('l'), 'label' => $text]); }
    public function alert(string $message, string $color = 'info'): self { return $this->add(['type' => 'ALERT', 'key' => $this->k('a'), 'message' => $message, 'color' => $color]); }
    public function badge(string $title, string $color = 'primary'): self { return $this->add(['type' => 'BADGE', 'key' => $this->k('bd'), 'title' => $title, 'color' => $color]); }
    public function spacer(int $px = 20): self { return $this->add(['type' => 'SPACER', 'key' => $this->k('sp'), 'width' => $px]); }

    // Inputs
    public function input(string $id, string $label = '', bool $required = false, string $dataType = 'STRING'): self {
        $el = ['type' => 'INPUT', 'key' => $this->k('i'), 'id' => $id, 'label' => $label];
        if ($required) $el['required'] = true; if ($dataType !== 'STRING') $el['dataType'] = $dataType;
        return $this->add($el);
    }
    public function textarea(string $id, string $label = '', int $rows = 3): self { return $this->add(['type' => 'TEXTAREA', 'key' => $this->k('ta'), 'id' => $id, 'label' => $label, 'rows' => $rows]); }
    public function checkbox(string $id, string $label = ''): self { return $this->add(['type' => 'CHECKBOX', 'key' => $this->k('cb'), 'id' => $id, 'label' => $label]); }
    public function select(string $id, string $label, array $pairs): self {
        $vals = array_map(fn($p) => ['id' => $p[0], 'displayName' => $p[1]], $pairs);
        return $this->add(['type' => 'SELECT', 'key' => $this->k('s'), 'id' => $id, 'label' => $label, 'values' => $vals]);
    }
    public function rating(string $id, string $label = ''): self { return $this->add(['type' => 'RATING', 'key' => $this->k('rt'), 'id' => $id, 'label' => $label]); }
    public function readonly(string $id, string $label = ''): self { return $this->add(['type' => 'READONLY_FIELD', 'key' => $this->k('ro'), 'id' => $id, 'label' => $label]); }

    // Actions
    public function button(string $id, string $title, string $color = 'primary', bool $isDefault = false): self {
        $el = ['type' => 'BUTTON', 'key' => $this->k('btn'), 'id' => $id, 'title' => $title, 'color' => $color];
        if ($isDefault) $el['default'] = true;
        $this->actions[] = $el; return $this;
    }

    // Shortcuts
    public function section(string $title, string $text): self { return $this->fieldset($title)->label($text)->end(); }
    public function feedbackForm(): self {
        return $this->fieldset('Your Details')
            ->input('name', 'Name', true)
            ->input('email', 'Email', true)
            ->textarea('message', 'Message', 5)
            ->end()
            ->button('send', 'Send', 'primary', true)
            ->button('cancel', 'Cancel', 'secondary');
    }

    // Output
    public function toArray(): array {
        while (count($this->stack) > 1) $this->pop();
        $ui = ['title' => $this->title, 'uid' => $this->uid, 'layout' => $this->layout, 'translations' => (object)[], 'userAccess' => ['cancel' => true]];
        if (!empty($this->actions)) $ui['actions'] = $this->actions;
        return ['ui' => $ui];
    }
    public function toJSON(): string { return json_encode($this->toArray(), JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE); }
    public function __toString(): string { return $this->toJSON(); }
}

File: DynamicLayout.swift
// DynamicLayout Swift API — generate layout JSON from Swift code.
//
// Single-file. Zero dependencies (Foundation is built-in).
// Works on Swift 5.5+, iOS 15+, macOS 12+.
//
// Example:
//   let page = Layout("My Page")
//   page.fieldset("Info").label("Hello from Swift!").endFieldset()
//   page.button("ok", "OK", "primary", isDefault: true)
//   print(page.end())

import Foundation

public class Layout {
    private var title: String
    private var uid: String
    private var layout: [[String: Any]] = []
    private var actions: [[String: Any]] = []
    private var stack: [[[String: Any]]] = []
    private var counter = 0

    public init(_ title: String) {
        self.title = title
        self.uid = "layout_\(Int(Date().timeIntervalSince1970 * 1000))"
        stack = [layout]
    }

    private func k(_ prefix: String) -> String { counter += 1; return "\(prefix)_\(counter)" }
    private var top: [[String: Any]] {
        get { stack[stack.count - 1] }
        set { stack[stack.count - 1] = newValue }
    }
    @discardableResult private func add(_ el: [String: Any]) -> Layout { top.append(el); return self }
    @discardableResult private func push(_ el: [String: Any]) -> Layout {
        var e = el; e["content"] = [[String: Any]](); add(e); stack.append(e["content"] as! [[String: Any]]); return self
    }
    @discardableResult private func pop() -> Layout { stack.removeLast(); return self }

    // Containers
    @discardableResult public func fieldset(_ title: String, collapsed: Bool? = nil) -> Layout {
        var el: [String: Any] = ["type": "FIELDSET", "key": k("fs"), "title": title]
        if let c = collapsed { el["collapsed"] = c }
        return push(el)
    }
    @discardableResult public func endFieldset() -> Layout { return pop() }
    @discardableResult public func row() -> Layout { return push(["type": "ROW", "key": k("r")]) }
    @discardableResult public func endRow() -> Layout { return pop() }
    @discardableResult public func col(xs: Int = 0, sm: Int = 0, md: Int = 0, lg: Int = 0, xl: Int = 0) -> Layout {
        var el: [String: Any] = ["type": "COL", "key": k("c")]
        var len = [String: Int]()
        if xs > 0 { len["xs"] = xs }; if sm > 0 { len["sm"] = sm }; if md > 0 { len["md"] = md }
        if lg > 0 { len["lg"] = lg }; if xl > 0 { len["xl"] = xl }
        if !len.isEmpty { el["length"] = len }
        return push(el)
    }
    @discardableResult public func endCol() -> Layout { return pop() }

    // Display
    @discardableResult public func label(_ text: String) -> Layout { return add(["type": "LABEL", "key": k("l"), "label": text]) }
    @discardableResult public func alert(_ message: String, color: String = "info") -> Layout { return add(["type": "ALERT", "key": k("a"), "message": message, "color": color]) }
    @discardableResult public func badge(_ title: String, color: String = "primary") -> Layout { return add(["type": "BADGE", "key": k("bd"), "title": title, "color": color]) }
    @discardableResult public func spacer(_ px: Int = 20) -> Layout { return add(["type": "SPACER", "key": k("sp"), "width": px]) }
    @discardableResult public func progress(_ pct: Int, label: String? = nil, color: String? = nil) -> Layout {
        var el: [String: Any] = ["type": "PROGRESS", "key": k("pr"), "progress": pct]
        if let l = label { el["label"] = l }; if let c = color { el["color"] = c }
        return add(el)
    }

    // Inputs
    @discardableResult public func input(_ id: String, label: String = "", required: Bool = false, dataType: String = "STRING") -> Layout {
        var el: [String: Any] = ["type": "INPUT", "key": k("i"), "id": id, "label": label]
        if required { el["required"] = true }; if dataType != "STRING" { el["dataType"] = dataType }
        return add(el)
    }
    @discardableResult public func textarea(_ id: String, label: String = "", rows: Int = 3) -> Layout { return add(["type": "TEXTAREA", "key": k("ta"), "id": id, "label": label, "rows": rows]) }
    @discardableResult public func checkbox(_ id: String, label: String = "") -> Layout { return add(["type": "CHECKBOX", "key": k("cb"), "id": id, "label": label]) }
    @discardableResult public func select(_ id: String, label: String, values: [(String, String)]) -> Layout {
        let vals = values.map { ["id": $0.0, "displayName": $0.1] as [String: Any] }
        return add(["type": "SELECT", "key": k("s"), "id": id, "label": label, "values": vals])
    }
    @discardableResult public func rating(_ id: String, label: String = "") -> Layout { return add(["type": "RATING", "key": k("rt"), "id": id, "label": label]) }
    @discardableResult public func readonly(_ id: String, label: String = "") -> Layout { return add(["type": "READONLY_FIELD", "key": k("ro"), "id": id, "label": label]) }

    // Actions
    @discardableResult public func button(_ id: String, _ title: String, _ color: String = "primary", isDefault: Bool = false) -> Layout {
        var el: [String: Any] = ["type": "BUTTON", "key": k("btn"), "id": id, "title": title, "color": color]
        if isDefault { el["default"] = true }
        actions.append(el); return self
    }

    // Shortcuts
    @discardableResult public func section(_ title: String, _ text: String) -> Layout { return fieldset(title).label(text).endFieldset() }
    @discardableResult public func feedbackForm() -> Layout {
        return fieldset("Your Details")
            .input("name", label: "Name", required: true)
            .input("email", label: "Email", required: true)
            .textarea("message", label: "Message", rows: 5)
            .endFieldset()
            .button("send", "Send", "primary", isDefault: true)
            .button("cancel", "Cancel", "secondary")
    }

    // Output
    public func end() -> String {
        while stack.count > 1 { pop() }
        var root: [String: Any] = ["title": title, "uid": uid, "layout": layout, "translations": [String: String](), "userAccess": ["cancel": true]]
        if !actions.isEmpty { root["actions"] = actions }
        let dict: [String: Any] = ["ui": root]
        if let data = try? JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted),
           let json = String(data: data, encoding: .utf8) { return json }
        return "{}"
    }
}
